// model.proto
syntax = "proto3";

//message ModelParams {
//  repeated float weights = 1;
//  int32 client_id = 2;
//}

enum ParamType {
  UNKNOWN = 0;
  ENCODER_WEIGHT = 1;
  ENCODER_BIAS = 2;
  CLASSIFIER_WEIGHT = 3;
  CLASSIFIER_BIAS = 4;
  FULL_MODEL = 5; // optional: 若是整個模型參數
}

message ModelParams {
  ParamType param_type = 1;        // 參數類型（新增）
  repeated float values = 2;       // 欲傳送的數值
  int32 client_id = 3;             // client 編號
}
message ServerResponse {
  string message = 1;
  bool update_successful = 2;
  int64 update_timestamp = 3;
}

service FederatedLearning {
  rpc UploadModelParams(ModelParams) returns (ServerResponse);
  rpc GetUpdateStatus(ModelParams) returns (ServerResponse);
}
